### PR TITLE
[WIP] repo level tools

### PR DESCRIPTION
**Pull request by enricosada**
fix https://github.com/fsprojects/Paket/issues/2919

to test it, an example in https://github.com/enricosada/paket-repotool-testing
aligned to this PR wip

- [X] `paket install` tools from nupkgs
- [X] `paket restore` tools from nupkgs
- [X] `paket add-tool FAKE` //install in repo: add to deps + install
- [x] global helper script to set repotool in PATH `repotool enable`/`disable` => `paketrt e` (or `enable`), `paketrt d` (or `disable`)
- [x] `paket add-tool FAKE -G`  //install globally
  - [X] first time: call `init` and `restore` if not found to install helpers and `paketg`
  - [X] install requested tool
  - [ ] other times: use `paketg add-tool FAKE` instead of directly use that `paket.dependencies`
  - is `paket add-tool -G` a good idea (as alias for `paketg add-tool`), or just drop it and ask directly to use `paketg`?
- [X] infer tools inside nupkg
    - [X] allow `tools/*.exe` (old directory structure)
    - [X] allow `tools/net*/*.exe` 
    - ~allow `tools/netcoreapp*/{pkgname}.dll`~
    - [X] allow `tools/netcoreapp*/{name}.dll` if `{name}.deps.json` exists in same dir
- [X] integration tests (`install`, `restore`)
- [x] add multi targeting tools support: 
     - [X] atm based on `PAKET_REPOTOOL_PREFFERED_RUNTIME` env var)
     - [x]  `net4*` (with mono on unix/osx) 
     - [x]  `netcoreapp2*` with dotnet, as fdd
- [X] create a `paket` wrapper in Main `bin` dir if a `.paket/paket.exe` exists 
- ~add helper script to add `bin` dir to PATH (windows). `paket-files\bin\paketrt e`~ preferred helper is global `paketrt`. but a paket command exists too to do it without `paketrt`
- ~add helper script to add `bin` dir to PATH (powershell). `. paket-files\bin\add_to_PATH.ps1`~ skipped powershell for this release
- ~add helper script to source `bin` dir to PATH (linux/osx). `. <(./paket-files/bin/paketrt e)`~ preferred helper is global `paketrt`. but a paket command exists too to do it without `paketrt`
- ~support configuration by xml in nupkg (for alias mostly)~ skipped until global tools stabilize
- [x] no relative dir, use absolute paths
- [ ] helper use written info in `paket-files/paket-repotools.csv` instead of read `paket.lock` (allow in a backward compatible way a different verison of paket to read installed tools info)
- [ ] if `paketrt` with empty args =>whoiam
- [ ] feedback of https://github.com/enricosada/paket-repotool-testing/#known-bugs-or-not-implemented-yet

### CODE REVIEW

**Comment by enricosada** [comment]
added another version for `paket` shell script, thx to @baronfel !

**Comment by enricosada** [comment]
Updated the demo repository after last time feedback, just clone and follow the README https://github.com/enricosada/paket-repotool-testing/

**Comment by enricosada** [comment]
Updated the demo repository (tagged as `v2.2` but master is always ok).
Just clone and follow the README https://github.com/enricosada/paket-repotool-testing/

File: src/Paket.Core/Paket.Core.fsproj [hunk 1]
```diff
@@ -2200,4 +2201,4 @@
   </Choose>
   <Import Project="..\..\packages\NETStandard.Library\build\NETStandard.Library.targets" Condition="Exists('..\..\packages\NETStandard.Library\build\NETStandard.Library.targets')" Label="Paket" />
   <Import Project="..\..\packages\NETStandard.Library\build\$(__paket__NETStandard_Library_targets).targets" Condition="Exists('..\..\packages\NETStandard.Library\build\$(__paket__NETStandard_Library_targets).targets')" Label="Paket" />
-</Project>
```
Comment (realvictorprm) [inline src/Paket.Core/Paket.Core.fsproj]: Why did this happen :(?

Comment (enricosada) [reply src/Paket.Core/Paket.Core.fsproj]: newline at EOF.

**Comment by matthid** [comment]
@enricosada Looks pretty ready to me. How about we just release this in a beta?

**Comment by enricosada** [comment]
rebased

**Comment by matthid** [comment]
This could be useful for ootb support for dotnet tools in FAKE, see https://github.com/fsharp/FAKE/issues/2169#issuecomment-524663384

@enricosada I might finish this (as it basically looks good to me), any objections regarding syntax?

For me, this also solves https://github.com/fsprojects/Paket/issues/3623 as I can just add paket and let fake restore paket (/cc @Krzysztof-Cieslak)

**Comment by enricosada** [comment]
Closing

some good things in the PR (lot of learning too for myself :D) but .net local tool are the way forward

**Pull request closed without merging**